File: packages/zig/src/stack.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

pub fn Stack(comptime T: type) type {
    return struct {
        items: std.ArrayList(T),
        allocator: Allocator,

        const Self = @This();

        pub fn init(allocator: Allocator) Self {
            return Self{ .items = .empty, .allocator = allocator };
        }

        pub fn deinit(self: *Self) void {
            self.items.deinit(self.allocator);
        }

        pub fn push(self: *Self, item: T) !void {
            try self.items.append(self.allocator, item);
        }

        pub fn pop(self: *Self) ?T {
            return self.items.pop();
        }

        pub fn peek(self: *Self) ?T {
            if (self.items.items.len == 0) return null;
            return self.items.items[self.items.items.len - 1];
        }
    };
}

pub fn isValidParentheses(allocator: Allocator, s: []const u8) !bool {
    var st = Stack(u8).init(allocator);
    defer st.deinit();

    for (s) |c| {
        switch (c) {
            '(', '[', '{' => try st.push(c),
            ')' => {
                if (st.pop()) |top| {
                    if (top != '(') return false;
                } else return false;
            },
            ']' => {
                if (st.pop()) |top| {
                    if (top != '[') return false;
                } else return false;
            },
            '}' => {
                if (st.pop()) |top| {
                    if (top != '{') return false;
                } else return false;
            },
            else => return false,
        }
    }
    return st.peek() == null;
}

File: packages/python/stack.py
class Stack:
    def __init__(self):
        self.items = []
        
    def push(self, item):
        self.items.append(item)
        
    def pop(self):
        if not self.is_empty():
            return self.items.pop()
        return None
        
    def is_empty(self):
        return len(self.items) == 0
        
    def peek(self):
        if not self.is_empty():
            return self.items[-1]
        return None

def is_valid_parentheses(s):
    stack = Stack()
    mapping = {")": "(", "}": "{", "]": "["}
    for char in s:
        if char in mapping:
            top_element = stack.pop() if not stack.is_empty() else '#'
            if mapping[char] != top_element:
                return False
        else:
            stack.push(char)
    return stack.is_empty()
